### PR TITLE
Add redis storage implementation in Auth

### PR DESCRIPTION
**Pull request by StringKe**
Closes: ?

### What are the changes and their implications?
Add redis storage implementation in Auth,

Supports both [ioredis](https://www.npmjs.com/package/ioredis) and [redis](https://www.npmjs.com/package/redis)

**Since Pipeline doesn't have a redis backend set up, I don't know how unit tests for redis can be added to the project and I need some help.**

change:
1. Separate the prisma implementation from the [auth-plugin](https://github.com/StringKe/blitz/tree/48b30eeb9c7e42585869326e0a00eb813af93f25/packages/blitz-auth/src/server/auth-plugin.ts) (move it to the [plugins](https://github.com/StringKe/blitz/tree/48b30eeb9c7e42585869326e0a00eb813af93f25/packages/blitz-auth/src/server/plugins) directory) and implement redis in a new directory ([plugins](https://github.com/StringKe/blitz/tree/48b30eeb9c7e42585869326e0a00eb813af93f25/packages/blitz-auth/src/server/plugins))

## Bug Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)

## Feature Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)
- [ ] Documentation added/updated (submit PR to [blitzjs.com repo `canary` branch](https://github.com/blitz-js/blitzjs.com/tree/canary))

### CODE REVIEW

File: packages/blitz-auth/package.json [hunk 1]
```diff
@@ -31,10 +31,12 @@
     "cookie-session": "2.0.0",
     "debug": "4.3.3",
     "http": "0.0.1-security",
+    "ioredis": "5.2.1",
```
Comment (beerose) [inline packages/blitz-auth/package.json:34]: I think we should find a solution where we avoid adding Redis to the dependencies. Currently, everyone using blitz-auth will end up having ioredis and redis in their deps even if they aren't using it.

**Review by beerose — COMMENTED** [review]
Thanks a lot for starting to work on that!

I left one comment with something up for a discussion

File: packages/blitz-auth/package.json (hunk 1 above)
Comment (StringKe) [reply packages/blitz-auth/package.json:34]: Both of these dependencies can be added to devDependencies , taking into account the typescript definitions.

Maybe it is possible to provide an abstraction layer up to the user concrete implementation to decide which one to use?

Comment (Zeko369) [reply packages/blitz-auth/package.json:34]: Maybe have `@blitzjs/auth-backend-redis`?

Comment (beerose) [reply packages/blitz-auth/package.json:34]: > Maybe have `@blitzjs/auth-backend-redis`?

Yeah, I was thinking about `@blitzjs/auth/redis` or something. Not sure yet what would be best here

Comment (flybayer) [reply packages/blitz-auth/package.json]: Can we keep it in `@blitzjs/auth` but move redis stuff to devDeps and have users install it themselves?

That's what we do with Prisma and Zod, and it works well

**Comment by lhocke** [comment]
@Zeko369 pointed me to this and I've been testing it out with ioredis, ran into a few issues when trying to actually create the session and add it to the blitz context.

1. ```expires``` needs to be converted to an integer before it can be passed into ```redis.set``` or it throws an error
2. ```getSession``` needs to return a ```Promise<SessionModel>``` to properly serialize a session on login or else it just creates an entry in the database that is inaccessible on the client.
3. attempting to resolve the parsed data from a redis query into blitz-context will throw ```date.getTime() is not a function``` unless ```expiresAt``` is first converted to a Date object (values are stored as strings in Redis)

I've written a potential replacement for [packages/blitz-auth-ioredis/src/index-server.ts](https://github.com/blitz-js/blitz/pull/3562/files#diff-45460f1d865d1d5a47e427e749e38b9ad61331844bbedc1b7d11adf1d36f21cc) that resolves these issues, that I can either include as a code snippet or fork and do a PR, whichever you'd prefer

File: packages/blitz-auth/src/server/plugins/auth-primsa.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import {PublicData, SessionConfigMethods, SessionModel} from "../../shared"
```
Comment (beerose) [inline packages/blitz-auth/src/server/plugins/auth-primsa.ts:1]: There's a type in the filename (primsa -> prisma)

File: packages/blitz-auth-redis/test/sample.test.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import {assert, expect, test} from "vitest"
```
Comment (beerose) [inline packages/blitz-auth-redis/test/sample.test.ts:1]: Are you planning to add tests for the auth-redis?

**Pull request closed without merging**